modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java

```diff
@@ -5524,27 +5526,85 @@ private void restorePartitionStates(
+                partIds.add(new LinkedList<>());
+
+            int cntr = 0;
+
+            // Group id -> (group ctx, completed partitions counter)
+            Map<Integer, T2<CacheGroupContext, AtomicInteger>> grps = new HashMap<>();
+
+            for (CacheGroupContext ctx : forGroups) {
+                grps.put(ctx.groupId(), new T2<>(ctx, new AtomicInteger()));
+
+                for (int i = 0; i < ctx.affinity().partitions(); i++) {
+                    partIds.get(cntr % poolSize).add(new GroupPartitionId(ctx.groupId(), i));
+
+                    cntr++;
+                }
+            }
+
+            CountDownLatch completionLatch = new CountDownLatch(cntr);
+
+            for (int i = 0; i < poolSize; i++) {
+                final int batchIdx = i;
+
+                sysPool.submit(() -> {
+                    List<GroupPartitionId> batch = partIds.get(batchIdx);
+
+                    Map<GroupPartitionId, Long> processed = new HashMap<>();
+
                     try {
-                        Map<Integer, Long> processed = grp.offheap().restorePartitionStates(partStates);
+                        for (GroupPartitionId grpPartId : batch) {
+                            T2<CacheGroupContext, AtomicInteger> grpTuple = grps.get(grpPartId.getGroupId());
+
+                            CacheGroupContext grpCtx = grpTuple.get1();
+
+                            try {
+                                long time = grpCtx.offheap().restoreStateOfPartition(grpPartId.getPartitionId(), partStates);
+
+                                processed.put(grpPartId, time);
 
-                        totalProcessed.addAndGet(processed.size());
+                                totalProcessed.incrementAndGet();
+                            }
+                            catch (IgniteCheckedException | RuntimeException | Error e) {
+                                U.error(log, "Failed to restore partition state for " +
+                                    "groupName=" + grpCtx.name() + " groupId=" + grpCtx.groupId(), e);
 
+                                IgniteCheckedException ex = e instanceof IgniteCheckedException
+                                    ? ((IgniteCheckedException)e)
+                                    : new IgniteCheckedException(e);
+
+                                if (!restoreStateError.compareAndSet(null, ex))
+                                    restoreStateError.get().addSuppressed(ex);
+                            }
+                            finally {
+                                completionLatch.countDown();
+
+                                AtomicInteger completedCntr = grpTuple.get2();
+
+                                if (completedCntr.incrementAndGet() == grpCtx.affinity().partitions())
+                                    grpCtx.offheap().confirmPartitionStatesRestored();
+                            }
+                        }
+                    }
+                    finally {
```

Comment:
This is not correct, since the buckets can be large and when outputting on line 5642, we will not see the actual information.